File: project files/server/src/controllers/adminController.js
const asyncHandler = require('../utils/asyncHandler');
const { getPagination } = require('../utils/pagination');
const orderRepository = require('../repositories/orderRepository');
const userRepository = require('../repositories/userRepository');
const inventoryService = require('../services/inventoryService');
const cacheService = require('../services/cacheService');
const stockMovementRepository = require('../repositories/stockMovementRepository');
const catalogAdminRepository = require('../repositories/catalogAdminRepository');
const bannerRepository = require('../repositories/bannerRepository');
const adminImportService = require('../services/adminImportService');
const AppError = require('../utils/AppError');
const { toCsv } = require('../utils/csv');
const catalogAdminService = require('../services/catalogAdminService');
const auditLogService = require('../services/auditLogService');
const homeFeaturedRepository = require('../repositories/homeFeaturedRepository');

const listOrders = asyncHandler(async (req, res) => {
  const { page, pageSize, skip } = getPagination(req.query);
  const query = {};
  if (req.query.status) query.orderStatus = req.query.status;
  if (req.query.paymentStatus) query.paymentStatus = req.query.paymentStatus;
  if (req.query.sellerId) query.sellerId = req.query.sellerId;
  if (req.query.orderId) query.orderId = req.query.orderId;
  if (req.query.fromDate || req.query.toDate) {
    query.createdAt = {};
    if (req.query.fromDate) query.createdAt.$gte = new Date(req.query.fromDate);
    if (req.query.toDate) query.createdAt.$lte = new Date(req.query.toDate);
  }
  const data = await orderRepository.listOrders({ query, page, pageSize, skip });
  res.status(200).json({ success: true, ...data });
});

const lowStockAlerts = asyncHandler(async (req, res) => {
  const cached = cacheService.getLowStockCache();
  if (cached.updatedAt && req.query.force !== 'true') {
    return res.status(200).json({
      success: true,
      data: cached.data,
      items: cached.data,
      total: cached.data.length,
      page: 1,
      pageSize: cached.data.length,
      totalPages: 1,
      updatedAt: cached.updatedAt,
      source: 'cache'
    });
  }
  const data = await inventoryService.scanLowStockVariants();
  cacheService.setLowStockCache(data);
  return res.status(200).json({
    success: true,
    data,
    items: data,
    total: data.length,
    page: 1,
    pageSize: data.length,
    totalPages: 1,
    updatedAt: new Date(),
    source: 'fresh'
  });
});

const listUsers = asyncHandler(async (req, res) => {
  const { page, pageSize, skip } = getPagination(req.query);
  const query = {};
  if (req.query.role) query.role = req.query.role;
  if (req.query.status) query.accountStatus = req.query.status;
  if (req.query.search) {
    query.$or = [
      { name: { $regex: req.query.search, $options: 'i' } },
      { email: { $regex: req.query.search, $options: 'i' } }
    ];
  }
  const data = await userRepository.listUsers({ query, page, pageSize, skip });
  return res.status(200).json({ success: true, ...data });
});

const blockUser = asyncHandler(async (req, res) => {
  await userRepository.setBlockState(req.params.id, {
    block: !!req.body.block,
    reason: req.body.reason || ''
  });
  res.status(200).json({ success: true, message: req.body.block ? 'User blocked' : 'User unblocked' });
});

const updateUserRole = asyncHandler(async (req, res) => {
  const existingUser = await userRepository.findById(req.params.id);
  if (!existingUser) throw new AppError('User not found', 404, 'USER_NOT_FOUND');

  const user = await userRepository.updateById(req.params.id, { role: req.body.role });
  if (!user) throw new AppError('User not found', 404, 'USER_NOT_FOUND');
  await auditLogService.createAuditLog({
    action: 'ROLE_CHANGE',
    performedBy: req.auth.userId,
    targetType: 'USER',
    targetId: req.params.id,
    previousValue: { role: existingUser.role },
    newValue: { role: user.role },
    req
  });
  res.status(200).json({
    success: true,
    data: {
      id: user._id,
      email: user.email,
      role: user.role
    }
  });
});

const dashboardMetrics = asyncHandler(async (req, res) => {
  const todayStart = new Date();
  todayStart.setHours(0, 0, 0, 0);
  const yesterdayStart = new Date(todayStart);
  yesterdayStart.setDate(yesterdayStart.getDate() - 1);

  const [todayOrders, yesterdayOrders, lowStock, activeUsers] = await Promise.all([
    orderRepository.getOrdersRevenueSummary({ fromDate: todayStart }),
    orderRepository.getOrdersRevenueSummary({ fromDate: yesterdayStart, toDate: todayStart }),
    inventoryService.scanLowStockVariants(),
    userRepository.countActiveUsers()
  ]);

  const todayRevenue = todayOrders.revenue;
  const yesterdayRevenue = yesterdayOrders.revenue;
  const revenueDeltaPct = yesterdayRevenue ? Number((((todayRevenue - yesterdayRevenue) / yesterdayRevenue) * 100).toFixed(2)) : 100;

  res.status(200).json({
    success: true,
    data: {
      todayRevenue,
      todayOrders: todayOrders.orders,
      activeUsers,
      lowStockCount: lowStock.length,
      revenueDeltaPct
    }
  });
});

const dashboardAnalytics = asyncHandler(async (req, res) => {
  const rangeDays = Number(req.query.rangeDays || 30);
  const allowedRange = [7, 30, 90].includes(rangeDays) ? rangeDays : 30;
  const fromDate = req.query.fromDate ? new Date(req.query.fromDate) : null;
  const toDate = req.query.toDate ? new Date(req.query.toDate) : null;

  const startDate = fromDate || (() => {
    const d = new Date();
    d.setHours(0, 0, 0, 0);
    d.setDate(d.getDate() - (allowedRange - 1));
    return d;
  })();
  const endDate = toDate || null;
  const orderMatch = { createdAt: { $gte: startDate } };
  if (endDate) {
    orderMatch.createdAt.$lte = endDate;
  }

  const [metricsRes, revenueSeries, ordersByStatus, topSellingProducts, lowStock, recentOrders] = await Promise.all([
    (async () => {
      const todayStart = new Date();
      todayStart.setHours(0, 0, 0, 0);
      const yesterdayStart = new Date(todayStart);
      yesterdayStart.setDate(yesterdayStart.getDate() - 1);

      const [todayOrders, yesterdayOrders, activeUsers] = await Promise.all([
        orderRepository.getOrdersRevenueSummary({ fromDate: todayStart }),
        orderRepository.getOrdersRevenueSummary({ fromDate: yesterdayStart, toDate: todayStart }),
        userRepository.countActiveUsers()
      ]);
      const todayRevenue = todayOrders.revenue;
      const yesterdayRevenue = yesterdayOrders.revenue;
      return {
        todayRevenue,
        todayOrders: todayOrders.orders,
        activeUsers,
        revenueDeltaPct: yesterdayRevenue ? Number((((todayRevenue - yesterdayRevenue) / yesterdayRevenue) * 100).toFixed(2)) : 100
      };
    })(),
    fromDate || toDate
      ? orderRepository.getRevenueSeriesByRange({ fromDate, toDate })
      : orderRepository.getRevenueSeriesByDay({ days: allowedRange }),
    orderRepository.getOrdersByStatus(orderMatch),
    orderRepository.getTopSellingProductsFromOrders(10, orderMatch),
    inventoryService.scanLowStockVariants(),
    orderRepository.listRecentOrders(10, orderMatch)
  ]);

  res.status(200).json({
    success: true,
    data: {
      metrics: { ...metricsRes, lowStockCount: lowStock.length },
      revenueSeries,
      ordersByStatus,
      topSellingProducts,
      lowStock,
      recentOrders
    }
  });
});

const listStockMovements = asyncHandler(async (req, res) => {
  const { page, pageSize, skip } = getPagination(req.query);
  const query = {};
  if (req.query.productId) query.productId = req.query.productId;
  if (req.query.variantId) query.variantId = req.query.variantId;
  if (req.query.reason) query.reason = req.query.reason;
  const data = await stockMovementRepository.listMovements({ query, page, pageSize, skip });
  res.status(200).json({ success: true, ...data });
});

const listPendingSellerProducts = asyncHandler(async (req, res) => {
  const { page, pageSize, skip } = getPagination(req.query);
  const data = await catalogAdminRepository.listPendingSellerProducts({ page, pageSize, skip });
  res.status(200).json({ success: true, ...data });
});

const reviewSellerProduct = asyncHandler(async (req, res) => {
  const product = await catalogAdminRepository.getProductById(req.params.id);
  if (!product) throw new AppError('Product not found', 404, 'PRODUCT_NOT_FOUND');
  const previous = {
    adminApproved: product.adminApproved,
    isActive: product.isActive,
    adminReviewNote: product.adminReviewNote
  };

  const action = req.body.action;
  const note = req.body.note || '';
  if (action === 'APPROVE') {
    product.adminApproved = true;
    product.isActive = true;
  } else {
    product.adminApproved = false;
    product.isActive = false;
  }
  product.adminReviewNote = note;
  await catalogAdminRepository.saveProduct(product);
  await auditLogService.createAuditLog({
    action: action === 'APPROVE' ? 'PRODUCT_APPROVE' : 'PRODUCT_REJECT',
    performedBy: req.auth.userId,
    targetType: 'PRODUCT',
    targetId: req.params.id,
    previousValue: previous,
    newValue: {
      adminApproved: product.adminApproved,
      isActive: product.isActive,
      adminReviewNote: product.adminReviewNote
    },
    req
  });
  res.status(200).json({ success: true, data: product.toObject() });
});

const listBanners = asyncHandler(async (req, res) => {
  const { page, pageSize, skip } = getPagination(req.query);
  const data = await bannerRepository.listAllBannersPaged({ page, pageSize, skip });
  res.status(200).json({ success: true, data: data.items, ...data });
});

const reorderBanners = asyncHandler(async (req, res) => {
  const data = await catalogAdminRepository.reorderBanners(req.body.bannerIds || []);
  await auditLogService.createAuditLog({
    action: 'BANNER_REORDER',
    performedBy: req.auth.userId,
    targetType: 'BANNER',
    targetId: 'bulk',
    previousValue: null,
    newValue: { bannerIds: req.body.bannerIds || [] },
    req
  });
  res.status(200).json({ success: true, data });
});

const createProductCsvUploadJob = asyncHandler(async (req, res) => {
  const job = await adminImportService.createProductCsvImportJob({
    csvContent: req.body.csvContent,
    actor: req.auth
  });
  res.status(202).json({ success: true, data: { jobId: job._id, status: job.status } });
});

const getProductCsvUploadJob = asyncHandler(async (req, res) => {
  const data = await adminImportService.getImportJobStatus(req.params.jobId);
  if (!data) throw new AppError('Import job not found', 404, 'IMPORT_JOB_NOT_FOUND');
  res.status(200).json({ success: true, data });
});

const productCsvTemplate = asyncHandler(async (_req, res) => {
  const template = adminImportService.getProductCsvTemplate();
  res.setHeader('Content-Type', 'text/csv');
  res.setHeader('Content-Disposition', 'attachment; filename="product_template.csv"');
  res.status(200).send(template);
});

const getCsvUploadFailureReport = asyncHandler(async (req, res) => {
  const data = await adminImportService.getImportJobStatus(req.params.jobId);
  if (!data) throw new AppError('Import job not found', 404, 'IMPORT_JOB_NOT_FOUND');
  const rows = (data.failureReport || []).map((item) => ({
    rowNumber: item.rowNumber,
    reason: item.reason
  }));
  const csv = toCsv(['rowNumber', 'reason'], rows);
  res.setHeader('Content-Type', 'text/csv');
  res.setHeader('Content-Disposition', `attachment; filename="import_failures_${req.params.jobId}.csv"`);
  res.status(200).send(csv);
});

const createStockCsvUploadJob = asyncHandler(async (req, res) => {
  const job = await adminImportService.createStockCsvImportJob({
    csvContent: req.body.csvContent,
    actor: req.auth
  });
  res.status(202).json({ success: true, data: { jobId: job._id, status: job.status } });
});

const stockCsvTemplate = asyncHandler(async (_req, res) => {
  const template = adminImportService.getStockCsvTemplate();
  res.setHeader('Content-Type', 'text/csv');
  res.setHeader('Content-Disposition', 'attachment; filename="stock_template.csv"');
  res.status(200).send(template);
});

const updateLowStockThreshold = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.updateLowStockThreshold({
    productId: req.body.productId,
    threshold: req.body.threshold
  });
  res.status(200).json({ success: true, data });
});

const listAuditLogs = asyncHandler(async (req, res) => {
  const { page, pageSize, skip } = getPagination(req.query);
  const data = await auditLogService.listAuditLogs({ page, pageSize, skip });
  res.status(200).json({ success: true, ...data });
});

const getHomeFeaturedConfig = asyncHandler(async (_req, res) => {
  const data = await homeFeaturedRepository.getConfig();
  res.status(200).json({ success: true, data });
});

const upsertHomeFeaturedConfig = asyncHandler(async (req, res) => {
  const data = await homeFeaturedRepository.upsertConfig(req.body.items || []);
  res.status(200).json({ success: true, data });
});

module.exports = {
  listOrders,
  lowStockAlerts,
  listUsers,
  blockUser,
  updateUserRole,
  dashboardMetrics,
  dashboardAnalytics,
  listStockMovements,
  listPendingSellerProducts,
  reviewSellerProduct,
  listBanners,
  reorderBanners,
  createProductCsvUploadJob,
  getProductCsvUploadJob,
  productCsvTemplate,
  getCsvUploadFailureReport,
  createStockCsvUploadJob,
  stockCsvTemplate,
  updateLowStockThreshold,
  listAuditLogs,
  getHomeFeaturedConfig,
  upsertHomeFeaturedConfig
};

File: spring_back/src/main/java/com/groceryapp/controller/AdminController.java
package com.groceryapp.controller;

import com.groceryapp.exception.AppException;
import com.groceryapp.model.Banner;
import com.groceryapp.model.ImportJob;
import com.groceryapp.model.Product;
import com.groceryapp.security.AuthContext;
import com.groceryapp.service.AdminImportService;
import com.groceryapp.service.AdminService;
import com.groceryapp.util.CsvUtil;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/admin")
public class AdminController {
    private final AdminService adminService;
    private final AdminImportService adminImportService;

    public AdminController(AdminService adminService, AdminImportService adminImportService) {
        this.adminService = adminService;
        this.adminImportService = adminImportService;
    }

    @GetMapping("/dashboard/metrics")
    public Map<String, Object> dashboardMetrics() {
        return Map.of("success", true, "data", adminService.dashboardMetrics());
    }

    @GetMapping("/dashboard/analytics")
    public Map<String, Object> dashboardAnalytics(@RequestParam Map<String, String> queryParams) {
        return Map.of("success", true, "data", adminService.dashboardAnalytics(queryParams));
    }

    @GetMapping("/orders")
    public Map<String, Object> listOrders(@RequestParam Map<String, String> queryParams) {
        return withSuccess(adminService.listOrders(queryParams));
    }

    @GetMapping("/inventory/low-stock")
    public Map<String, Object> lowStockAlerts(@RequestParam(required = false) String force) {
        Map<String, Object> data = adminService.lowStockAlerts("true".equalsIgnoreCase(force));
        return Map.of(
                "success", true,
                "data", data.get("data"),
                "items", data.get("items"),
                "total", data.get("total"),
                "page", data.get("page"),
                "pageSize", data.get("pageSize"),
                "totalPages", data.get("totalPages"),
                "updatedAt", data.get("updatedAt"),
                "source", data.get("source")
        );
    }

    @GetMapping("/inventory/movements")
    public Map<String, Object> listStockMovements(@RequestParam Map<String, String> queryParams) {
        return withSuccess(adminService.listStockMovements(queryParams));
    }

    @GetMapping("/users")
    public Map<String, Object> listUsers(@RequestParam Map<String, String> queryParams) {
        return withSuccess(adminService.listUsers(queryParams));
    }

    @GetMapping("/audit-logs")
    public Map<String, Object> listAuditLogs(@RequestParam Map<String, String> queryParams) {
        return withSuccess(adminService.listAuditLogs(queryParams));
    }

    @PatchMapping("/users/{id}/block")
    public Map<String, Object> blockUser(@PathVariable String id, @RequestBody Map<String, Object> body) {
        String message = adminService.blockUser(id, bool(body.get("block")), str(body.get("reason")));
        return Map.of("success", true, "message", message);
    }

    @PatchMapping("/users/{id}/role")
    public Map<String, Object> updateUserRole(@PathVariable String id, @RequestBody Map<String, Object> body, HttpServletRequest request) {
        Map<String, Object> data = adminService.updateUserRole(id, str(body.get("role")), AuthContext.current().getUserId(), request);
        return Map.of("success", true, "data", data);
    }

    @GetMapping("/products/csv-template")
    public ResponseEntity<String> productCsvTemplate() {
        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType("text/csv"))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"product_template.csv\"")
                .body(adminImportService.getProductCsvTemplate());
    }

    @PostMapping("/products/bulk-upload")
    public ResponseEntity<Map<String, Object>> createProductCsvUploadJob(@RequestBody Map<String, Object> body) {
        ImportJob job = adminImportService.createProductCsvImportJob(str(body.get("csvContent")), AuthContext.current().getUserId());
        return ResponseEntity.status(202).body(Map.of(
                "success", true,
                "data", Map.of("jobId", job.getId(), "status", job.getStatus())
        ));
    }

    @GetMapping("/products/bulk-upload/{jobId}")
    public Map<String, Object> getProductCsvUploadJob(@PathVariable String jobId) {
        ImportJob data = adminImportService.getImportJobStatus(jobId);
        if (data == null) throw new AppException("Import job not found", 404, "IMPORT_JOB_NOT_FOUND");
        return Map.of("success", true, "data", data);
    }

    @GetMapping("/products/bulk-upload/{jobId}/failures")
    public ResponseEntity<String> getCsvUploadFailureReport(@PathVariable String jobId) {
        ImportJob data = adminImportService.getImportJobStatus(jobId);
        if (data == null) throw new AppException("Import job not found", 404, "IMPORT_JOB_NOT_FOUND");
        List<Map<String, Object>> rows = (data.getFailureReport() == null ? List.<com.groceryapp.model.ImportJob.Failure>of() : data.getFailureReport())
                .stream()
                .map(item -> {
                    Map<String, Object> row = new HashMap<>();
                    row.put("rowNumber", item.getRowNumber());
                    row.put("reason", item.getReason());
                    return row;
                })
                .toList();
        String csv = CsvUtil.toCsv(List.of("rowNumber", "reason"), rows);
        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType("text/csv"))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"import_failures_" + jobId + ".csv\"")
                .body(csv);
    }

    @GetMapping("/inventory/stock-csv-template")
    public ResponseEntity<String> stockCsvTemplate() {
        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType("text/csv"))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"stock_template.csv\"")
                .body(adminImportService.getStockCsvTemplate());
    }

    @PostMapping("/inventory/bulk-stock-upload")
    public ResponseEntity<Map<String, Object>> createStockCsvUploadJob(@RequestBody Map<String, Object> body) {
        ImportJob job = adminImportService.createStockCsvImportJob(str(body.get("csvContent")), AuthContext.current().getUserId());
        return ResponseEntity.status(202).body(Map.of(
                "success", true,
                "data", Map.of("jobId", job.getId(), "status", job.getStatus())
        ));
    }

    @PatchMapping("/inventory/threshold")
    public Map<String, Object> updateLowStockThreshold(@RequestBody Map<String, Object> body) {
        Product data = adminService.updateLowStockThreshold(str(body.get("productId")), intVal(body.get("threshold"), 0));
        return Map.of("success", true, "data", data);
    }

    @GetMapping("/products/pending-approval")
    public Map<String, Object> listPendingSellerProducts(@RequestParam Map<String, String> queryParams) {
        return withSuccess(adminService.listPendingSellerProducts(queryParams));
    }

    @PatchMapping("/products/{id}/review")
    public Map<String, Object> reviewSellerProduct(@PathVariable String id, @RequestBody Map<String, Object> body, HttpServletRequest request) {
        Product data = adminService.reviewSellerProduct(
                id,
                str(body.get("action")),
                str(body.get("note")),
                AuthContext.current().getUserId(),
                request
        );
        return Map.of("success", true, "data", data);
    }

    @GetMapping("/banners")
    public Map<String, Object> listBanners(@RequestParam Map<String, String> queryParams) {
        return withSuccess(adminService.listBanners(queryParams));
    }

    @PatchMapping("/banners/reorder")
    public Map<String, Object> reorderBanners(@RequestBody Map<String, Object> body, HttpServletRequest request) {
        List<String> bannerIds = castStringList(body.get("bannerIds"));
        List<Banner> data = adminService.reorderBanners(bannerIds, AuthContext.current().getUserId(), request);
        return Map.of("success", true, "data", data);
    }

    @GetMapping("/home-featured")
    public Map<String, Object> getHomeFeaturedConfig() {
        return Map.of("success", true, "data", adminService.getHomeFeaturedConfig());
    }

    @PostMapping("/home-featured/upsert")
    public Map<String, Object> upsertHomeFeaturedConfig(@RequestBody Map<String, Object> body) {
        return Map.of("success", true, "data", adminService.upsertHomeFeaturedConfig(castMapList(body.get("items"))));
    }

    private Map<String, Object> withSuccess(Map<String, Object> data) {
        Map<String, Object> out = new HashMap<>();
        out.put("success", true);
        out.putAll(data);
        return out;
    }

    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> castMapList(Object value) {
        if (value instanceof List<?> list) return (List<Map<String, Object>>) list;
        return List.of();
    }

    private List<String> castStringList(Object value) {
        if (!(value instanceof List<?> rows)) return List.of();
        List<String> out = new ArrayList<>();
        for (Object row : rows) {
            String v = str(row);
            if (!v.isBlank()) out.add(v);
        }
        return out;
    }

    private String str(Object v) {
        return v == null ? "" : String.valueOf(v).trim();
    }

    private int intVal(Object v, int def) {
        try {
            return Integer.parseInt(str(v));
        } catch (Exception ex) {
            return def;
        }
    }

    private boolean bool(Object v) {
        if (v instanceof Boolean b) return b;
        return "true".equalsIgnoreCase(str(v)) || "1".equals(str(v));
    }
}
